lib/github/ldap/member_search/recursive.rb

```diff
@@ -34,73 +34,72 @@ def initialize(ldap, options = {})
         #
         # Returns Array of Net::LDAP::Entry objects.
         def perform(group)
-          found = Hash.new
-
-          # find members (N queries)
-          entries = member_entries(group)
-          return [] if entries.empty?
-
-          # track found entries
-          entries.each do |entry|
-            found[entry.dn] = entry
+          found    = Hash.new
+          searched = []
+          entries  = []
+
+          # if this is a posixGroup, return members immediately (no nesting)
+          uids = member_uids(group)
+          return entries_by_uid(uids) if uids.any?
+
+          # track group
+          searched << group.dn
+          found[group.dn] = group
+
+          # pull out base group's member DNs
+          dns = member_dns(group)
+
+          # search for base group's subgroups
+          filter = ALL_GROUPS_FILTER
+          groups = dns.each_with_object([]) do |dn, groups|
+            groups.concat ldap.search(base: dn, scope: Net::LDAP::SearchScope_BaseObject, attributes: attrs, filter: filter)
+            searched << dn
           end
 
-          # descend to `depth` levels, at most
-          depth.times do |n|
-            # find every (new, unique) member entry
-            depth_subentries = entries.each_with_object([]) do |entry, depth_entries|
-              submembers = entry["member"]
+          # track found groups
+          groups.each { |g| found[g.dn] = g }
 
-              # skip any members we've already found
-              submembers.reject! { |dn| found.key?(dn) }
+          # recursively find subgroups
+          unless groups.empty?
+            depth.times do |n|
+              # pull out subgroups' member DNs to search through
+              sub_dns = groups.each_with_object([]) do |subgroup, sub_dns|
+                sub_dns.concat member_dns(subgroup)
+              end
 
-              # find members of subgroup, including subgroups (N queries)
-              subentries = member_entries(entry)
-              next if subentries.empty?
+              # give up if there's nothing else to search for
+              break if sub_dns.empty?
 
-              # track found subentries
-              subentries.each { |entry| found[entry.dn] = entry }
+              # filter out if already searched for
+              sub_dns.reject! { |dn| searched.include?(dn) }
```

Comment:
I'm not sure whether ruby's Set is implemented with Array's, but depending on how large the Array grows, tracking may be faster with a Set/Hash here.